chore(js-loading): Make JS_URL be used at runtime for js loading

hazzadous opened the pull request:
## Changes

Previously we would bake the JS_URL in at build time. Instead this changes such that we can change this at runtime by specifying a `JS_URL` env variable.

The driver for this is that, in setting up a [staging environment for cloud](https://github.com/PostHog/posthog-cloud-infra/pull/19), we want to make it as close to a self hosted version as reasonable, e.g. using the same docker image. We're getting away with this at the moment by [hardcoding JS_URL in our production docker image](https://github.com/PostHog/posthog-cloud/blob/master/prod.web.Dockerfile#L3). Handling as separate images something that we want to vary at runtime isn't ideal.

## How did you test this code?

<!-- If the answer is manually, please include a quick step-by-step on how to test this PR. -->

I'm relying on the E2E tests here, but this isn't representative of the production ECS deployment so there is some risk here.

Commit:
chore(js-loading): Make JS_URL be used at runtime for js loading

Previously we would bake the JS_URL in at build time. Instead this
changes such that we can change this at runtime by specifying a `JS_URL`
env variable.

Commit:
comment on css loader snippet

Commit:
Default js_url to /

guidoiaquinti commented on frontend/utils.mjs:
```diff
@@ -92,17 +97,25 @@ export function copyIndexHtml(
+    const cssLoader = `
+        const link = document.createElement("link");
+        link.rel = "stylesheet";
+        link.href = window.JS_URL + "static/${cssFile}"
```
Qq: should `static/` be hardcoded? It was before so I think it is fine for now

hazzadous replied:
Yeah my reason for leaving is just to not change anything other than the required. I'd rather get staging a reproduction of production, get that workflow going, then we can consider how we can improve.

guidoiaquinti reviewed (approved):
LGTM. 

Non-blocking: can we rename `JS_URL` to something more descriptive? We don't load only JS stuff there. Django uses `STATIC_URL`, `STATIC_ROOT` (see [docs](https://docs.djangoproject.com/en/4.0/ref/settings/#static-url)).

hazzadous commented:
> LGTM.
> 
> Non-blocking: can we rename `JS_URL` to something more descriptive? We don't load only JS stuff there. Django uses `STATIC_URL`, `STATIC_ROOT` (see [docs](https://docs.djangoproject.com/en/4.0/ref/settings/#static-url)).

Sure, something like `FRONTEND_JS_BUNDLE_URL`?

guidoiaquinti commented:
> STATIC_URL

`STATIC_HOST`? Then we can define `STATIC_URL`=`STATIC_HOST` + `/static` to keep the consistency with Django.

hazzadous commented:
> `STATIC_HOST`? Then we can define `STATIC_URL`=`STATIC_HOST` + `/static` to keep the consistency with Django.

This is specifically the url for the JS bundle built by esbuild for the SPA though, rather than what would be picked up by django-admin.py collectstatic.

I'll leave as is for now and follow up with a change

Commit:
Leave leading / when concatenating to JS_URL

I'd accidentally left this out before

mariusandra commented on frontend/utils.mjs:
```diff
@@ -92,17 +97,25 @@ export function copyIndexHtml(
                     ${scriptCode}
                     ${Object.keys(chunks).length > 0 ? chunkCode : ''}
                 </script>
-                ${cssLinkTag}
```
Some context below regarding the order of things.

The CSS link tag (`<link rel="stylesheet" ... />`) that was there before blocks execution of the rest of the page until it finishes loading. When we had this tag before the `<script>` block, the initial page load for the user was sluggish because of a white screen. The page would load the CSS and only then unblock and start loading scripts. Changing the order of things helped dramatically.

It seems like dynamically injecting a CSS link tag does not make it block the same way actually having the tag in the DOM does. Both asset types start loading simultaneously:

![2022-01-27 13 55 54](https://user-images.githubusercontent.com/53387/151363143-dbd748eb-5a2d-424f-a7ec-c3812f9880f6.gif)

However, we now have the danger of running into what's called "the flash of unstyled content". We can end up in a scenario where the CSS is still loading, but all the JS has finished downloading and running... and the app has rendered, just without the CSS. 

I don't know how common this would be, but we should watch out.

hazzadous replied:
Yeah I'm not exactly sure on what the order of loading is. I've added a comment regarding this being a risk. Do you think there is anything else to do here?

mariusandra reviewed (changes requested):
Seems legit, but one issue with the defaults.

Without having a JS_URL set for myself, my dev env doesn't work after opening this PR and  running `yarn start` like normally:

![image](https://user-images.githubusercontent.com/53387/151362262-bd32b3b9-3857-48d1-83d6-8c9cee31eb59.png)
![image](https://user-images.githubusercontent.com/53387/151362604-e33f701e-362a-49d4-9a83-2dbd0b62c356.png)

I had to manually specify a JS_URL to get it to run. Changing [the default here](https://github.com/PostHog/posthog/blob/c093b593ac88a550504154520e06969903e16afc/posthog/settings/__init__.py#L55) to be slashless would be a good start, yet I think some de-slashing could would be better. Being loose with the input/envs we accept would reduce some support load, and make it a less frustrating experience for the users.

I also left two other thoughts inline, but non blocking.

hazzadous commented:
Great, thanks @mariusandra  I'll work through these points in a bit 🙏

Commit:
Remove trailing forward-slash from `JS_URL` django setting

Prior to the change to make JS_URL applicable at runtime, this variable
had no effect, hence the railing slash wasn't an issue.

This change fixes the frontend when running with debug true, i.e. when
we should be using the dev server for serving the frontend

Commit:
remove trailing slashes from JS_URL

Commit:
avoid unexpected concat behaviour

hazzadous commented on frontend/utils.mjs:
```diff
@@ -69,7 +74,7 @@ export function copyIndexHtml(
         window.ESBUILD_LOAD_SCRIPT = async function (file) {
             try {
-                await import('${useJsURL ? jsURL : ''}/static/' + file)
+                await import(window.JS_URL + '/static/' + file)
```
Yup I was assuming that my head.html default to empty string would be in place, but I guess that is no guarantee

Commit:
add some comments around style sheet loading

mariusandra commented on frontend/utils.mjs:
```diff
@@ -92,17 +97,31 @@ export function copyIndexHtml(
+    const cssLoader = `
+        const link = document.createElement("link");
+        link.rel = "stylesheet";
+        link.href = window.JS_URL + "/static/${cssFile}"
```
Also `(window.JS_URL || '')`

hazzadous replied:
oops 😳

mariusandra reviewed (approved):
Found one more small thing. Giving the green checkmark already, as the rest looks good.

However, organizationally, I'd suggest splitting the change in the default env in `task-definition.web.json` into a separate PR. Knowing what a wild west frontend coding can be, there's about a 80% chance some unexpected CORS error or missing header that messes with cross domain cookies will take the site down and we'll need to quickly revert something 😅 

![image](https://user-images.githubusercontent.com/53387/151382671-e371e1cd-56ac-40e8-9eab-90a49edbfb64.png)

Commit:
explicitly handle JS_URL not set for css as well

Commit:
chore(static-assets): set JS_URL for ECS

This is the breakaway change
[here](https://github.com/PostHog/posthog/pull/8299#pullrequestreview-864974223)

This will need to be merged in before the other change, as otherwise we
will end up using the default of /static/ to serve up the frontend in
production.
